### PR TITLE
feat: support for assigning an id to each query and pass it to ts-store

### LINKED ISSUE #291: support for assigning an id to each query and pass it to the store
support for assigning an id to each query and pass it to the store

### PR DESCRIPTION
**Pull request by wtsclwq**
### What problem does this PR solve?

Issue Number: fix #291

### What is changed and how it works?

Please describe how it works

Every ts-sql node can assgin different query id for every query, the id range offsets are maintained by ts-meta node in raft

image

Reference discussion #285

1. ts-sql：

   When ts-sql starts, it interacts with ts-meta by `ExecuteRequestMessage`  to register a query id start offset  for `TaskManager` to assgin query id

2. ts-meta：

   Meta node store a `QueryIDInit`  in its fsm.

   ```go
   const QueryIDSpan 100000
   
   type Data struct {
        ......
       // Query ID range segment allocated by all sql nodes
       QueryIDInit map[string]uint64 // {"127.0.0.1:8086": 0, "127.0.0.2:8086": 10w, "127.0.0.3:8086": 20w}, span is QueryIDSpan
   }
   ```

3. ts-store:

   Get query id and sql.String() from `RemoteQuery`

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B
- [x] Test cases to be added
- [ ] No code

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: open_src/influx/query/task_manager.go [hunk 1]
```diff
@@ -329,3 +339,58 @@ func (t *TaskManager) Close() error {
...
+// reuseQueryIDFromBeginning Used to find available qid from the beginning of (queryIDOffset,queryIDUpperLimit)
+func (t *TaskManager) reuseQueryIDFromBeginning() (uint64, error) {
+	// find a free query id from queryIDOffset to queryIDUpperLimit
+	for resID := t.queryIDOffset + 1; resID < t.queryIDUpperLimit; resID++ {
+		if _, ok := t.queries[resID]; !ok {
+			return resID, nil
+		}
+	}
+	return 0, ErrQueryIDExhausted
+}
```
Comment (shilinlee) [inline open_src/influx/query/task_manager.go:396]:
```suggestion
func (t *TaskManager) AssignQueryID() uint64 {
	atomic.CompareAndSwapUint64(&t.nextID, t.queryIDOffset+IdSpan, t.queryIDOffset)
	qid := atomic.AddUint64(&t.nextID, 1)
	return qid
}
```

File: open_src/influx/query/task_manager.go [hunk 2]
```diff
@@ -329,3 +339,58 @@ func (t *TaskManager) Close() error {
 func (t *TaskManager) Statistics(buffer []byte) ([]byte, error) {
 	return nil, nil
 }
+
+func (t *TaskManager) InitQueryIDByOffset(offset uint64) {
+	t.nextID = offset + 1
```
Comment (shilinlee) [inline open_src/influx/query/task_manager.go:344]:
```suggestion
	t.nextID = offset
```

**Review by shilinlee — CHANGES_REQUESTED** [review]

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: open_src/influx/query/task_manager.go
Comment (shilinlee) [inline open_src/influx/query/task_manager.go:1]:
```go
func (t *TaskManager) KillQuery(qid uint64) error {
	t.mu.RLock()
	query := t.queries[qid]
	t.mu.RUnlock()
```

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: open_src/influx/meta/proto/meta.proto
Comment (shilinlee) [inline open_src/influx/meta/proto/meta.proto:1]: You can write a README under the folder?

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: lib/metaclient/meta_client.go [hunk 3]
```diff
@@ -3021,6 +3023,26 @@ func (c *Client) Suicide(err error) {
...
 }
 
+// RegisterQueryIDOffset send a register rpc to ts-meta，request a query id offset
+func (c *Client) RegisterQueryIDOffset(host string) error {
+	if _, ok := c.cacheData.QueryIDInit[host]; ok {
+		c.logger.Warn("current host has already registered in ts-meta")
+		return nil
+	}
+	cmd := &proto2.RegisterQueryIDOffsetCommand{Host: proto.String(host)}
+	err := c.retryUntilExec(proto2.Command_RegisterQueryIDOffsetCommand, proto2.E_RegisterQueryIDOffsetCommand_Command, cmd)
```
Comment (shilinlee) [inline lib/metaclient/meta_client.go:3033]:
```suggestion
	err := c.retryExec(proto2.Command_RegisterQueryIDOffsetCommand, proto2.E_RegisterQueryIDOffsetCommand_Command, cmd)
```

File: app/ts-meta/meta/store_fsm.go [hunk 4]
```diff
@@ -704,3 +706,10 @@ func (fsm *storeFSM) applyVerifyDataNodeCommand(cmd *proto2.Command) interface{}
 
 	return nil
 }
+
+func (fsm *storeFSM) applyRegisterQueryIDOffsetCommand(cmd *proto2.Command) interface{} {
+	ext, _ := proto.GetExtension(cmd, proto2.E_RegisterQueryIDOffsetCommand_Command)
+	v := ext.(*proto2.RegisterQueryIDOffsetCommand)
```
Comment (shilinlee) [inline app/ts-meta/meta/store_fsm.go:712]:
```suggestion
	v, ok := ext.(*proto2.RegisterQueryIDOffsetCommand)
	if !ok {
		panic(fmt.Errorf("%s is not a RegisterQueryIDOffsetCommand", ext))
	}
```

File: open_src/influx/meta/data.go [hunk 5]
```diff
@@ -2769,6 +2805,24 @@ func (data *Data) checkDDLConflict(e *proto2.MigrateEventInfo) error {
...
+	}
+
+	currentAssignedNum := len(data.QueryIDInit)
+	if currentAssignedNum >= MaxAssignableNum {
+		return errno.NewError(errno.QueryIDOverflow)
+	}
+
+	nextOffset := uint64(currentAssignedNum * QueryIDSpan)
+
+	data.QueryIDInit[host] = nextOffset
```
Comment (shilinlee) [inline open_src/influx/meta/data.go:2821]: Should add a judgment whether the `host` already exists in `QueryIDInit`?

File: open_src/influx/meta/proto/meta.proto
Comment (wtsclwq) [reply open_src/influx/meta/proto/meta.proto:1]: About what?
1. Instructions for the **commands**?
2. How to modify `.proto` to update `.go`?

File: open_src/influx/meta/data.go [hunk 6]
```diff
@@ -2769,6 +2805,24 @@ func (data *Data) checkDDLConflict(e *proto2.MigrateEventInfo) error {
...
 
+// RegisterQueryIDOffset register the mapping relationship between its host and query id offset for ts-sql
+func (data *Data) RegisterQueryIDOffset(host string) error {
+	if data.QueryIDInit == nil {
+		data.QueryIDInit = make(map[string]uint64)
+	}
+
+	currentAssignedNum := len(data.QueryIDInit)
+	if currentAssignedNum >= MaxAssignableNum {
+		return errno.NewError(errno.QueryIDOverflow)
```
Comment (shilinlee) [inline open_src/influx/meta/data.go:2816]: `MaxAssignableNum = 1.8446744 * 10^14`
It is impossible to have so many SQLs.
I suggest not adding this `if`

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: open_src/influx/query/executor.go [hunk 7]
```diff
@@ -51,6 +51,9 @@ var (
 
 	// ErrAlreadyKilled is returned when attempting to kill a query that has already been killed.
 	ErrAlreadyKilled = errors.New("already killed")
+
+	// ErrQueryIDExhausted is returned when task manager assign a new query id that overflow range
+	ErrQueryIDExhausted = errors.New("query id exhausted")
```
Comment (shilinlee) [inline open_src/influx/query/executor.go:56]: Do we need this error?

**Review by shilinlee — CHANGES_REQUESTED** [review]

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: lib/metaclient/meta_client_test.go
Comment (shilinlee) [inline lib/metaclient/meta_client_test.go:1]: Too long to run this test case. It should less than 5s.

**Review by shilinlee — CHANGES_REQUESTED** [review]

**Review by shilinlee — CHANGES_REQUESTED** [review]

**Review by shilinlee — CHANGES_REQUESTED** [review]

**Review by shilinlee — APPROVED** [review]

**Pull request closed without merging**